crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,406 @@
+use core::{convert::Infallible, ops::AddAssign};
+
+use bitcoin::{OutPoint, Script, Transaction, TxOut};
+
+use crate::{
+    keychain::Balance,
+    sparse_chain::ChainPosition,
+    tx_graph::{Additions, TxGraph, TxNode},
+    BlockAnchor, ChainOracle, FullTxOut, ObservedIn, TxIndex,
+};
+
+/// An outwards-facing view of a transaction that is part of the *best chain*'s history.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxInChain<'a, T, A> {
+    /// Where the transaction is observed (in a block or in mempool).
+    pub observed_in: ObservedIn<&'a A>,
+    /// The transaction with anchors and last seen timestamp.
+    pub tx: TxNode<'a, T, A>,
+}
+
+/// A structure that represents changes to an [`IndexedTxGraph`].
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "A: Ord + serde::Deserialize<'de>, IA: serde::Deserialize<'de>",
+            serialize = "A: Ord + serde::Serialize, IA: serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct IndexedAdditions<A, IA> {
+    /// [`TxGraph`] additions.
+    pub graph_additions: Additions<A>,
+    /// [`TxIndex`] additions.
+    pub index_additions: IA,
+    /// Last block height witnessed (if any).
+    pub last_height: Option<u32>,
+}
+
+impl<A, IA: Default> Default for IndexedAdditions<A, IA> {
+    fn default() -> Self {
+        Self {
+            graph_additions: Default::default(),
+            index_additions: Default::default(),
+            last_height: None,
+        }
+    }
+}
+
+impl<A: BlockAnchor, IA: AddAssign> AddAssign for IndexedAdditions<A, IA> {
+    fn add_assign(&mut self, rhs: Self) {
+        let Self {
+            graph_additions,
+            index_additions: index_delta,
+            last_height,
+        } = rhs;
+        self.graph_additions.append(graph_additions);
+        self.index_additions += index_delta;
+        if self.last_height < last_height {
+            let last_height =
+                last_height.expect("must exist as it is larger than self.last_height");
+            self.last_height.replace(last_height);
+        }
+    }
+}
+
+pub struct IndexedTxGraph<A, I> {
+    /// Transaction index.
+    pub index: I,
```

Comment:
Is this a Transaction index or a SpkIndex? Are they same thing? 

I think it's better to name them separately. A lot of chain position indexing is also going around the crates. And a "transaction index" doesn't' explicitly mean it's related to the spk indexing of the keychains.